<h1><center><font size="6">Data Cleaning</font></center></h1>

# Outline
- [Import necessary Libraries](#1)
- [Introduction](#2)
- [Data Cleaning](#3)
- [Conclusions](#4)
- [References](#5)

<a id='1'></a>
# Import necessary Libraries

In [1]:
# Import necessary libraries and print their versions
from platform import python_version
print(f"Python: {python_version()}")
import numpy as np
print(f"Numpy: {np.__version__}")
import pandas as pd
print(f"Pandas: {pd.__version__}")

Python: 3.9.16
Numpy: 1.24.1
Pandas: 1.5.2


<a id='2'></a>
# Introduction
In this notebook we will prepare the data obtained from the web scraping for further processing. This involves removing irrelevant data and converting the data into a desired format.

<a id='3'></a>
# Data Cleaning
## Load Data

In [2]:
# Load data
df = pd.read_csv("../web_scraping/game_sales.csv", encoding="utf-8")
# Take a quick look if the import was successful
df.head()

Rank              Name Platform    Year             Genre  Critic_Score  \
0     1            Tetris   Series  1989.0            Puzzle           NaN   
1     2           Pokemon   Series  1998.0      Role-Playing           NaN   
2     3      Call of Duty   Series  2003.0           Shooter           NaN   
3     4       Super Mario   Series  1983.0          Platform           NaN   
4     5  Grand Theft Auto   Series  1998.0  Action-Adventure           NaN   

   User_Score         Publisher          Developer  NA_Sales  PAL_Sales  \
0         NaN        Nintendo    Alexey Pajitnov         NaN        NaN   
1         NaN        Nintendo         Game Freak         NaN        NaN   
2         NaN      Activision      Infinity Ward         NaN        NaN   
3         NaN        Nintendo           Nintendo         NaN        NaN   
4         NaN  Rockstar Games     Rockstar North         NaN        NaN   

   JP_Sales  Other_Sales  Total_Sales  
0       NaN          NaN          NaN  
1       NaN          NaN          NaN  
2       NaN          NaN          NaN  
3       NaN          NaN          NaN  
4       NaN          NaN          NaN

## Dealing with missing values

In [3]:
# See the column data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          25000 non-null  int64  
 1   Name          25000 non-null  object 
 2   Platform      25000 non-null  object 
 3   Year          24341 non-null  float64
 4   Genre         24994 non-null  object 
 5   Critic_Score  4987 non-null   float64
 6   User_Score    287 non-null    float64
 7   Publisher     25000 non-null  object 
 8   Developer     25000 non-null  object 
 9   NA_Sales      13660 non-null  float64
 10  PAL_Sales     14024 non-null  float64
 11  JP_Sales      7675 non-null   float64
 12  Other_Sales   16369 non-null  float64
 13  Total_Sales   20306 non-null  float64
dtypes: float64(8), int64(1), object(5)
memory usage: 2.7+ MB


In [4]:
# Check for missing values
df.isnull().sum()

Rank                0
Name                0
Platform            0
Year              659
Genre               6
Critic_Score    20013
User_Score      24713
Publisher           0
Developer           0
NA_Sales        11340
PAL_Sales       10976
JP_Sales        17325
Other_Sales      8631
Total_Sales      4694
dtype: int64

In [5]:
# Remove column 'User_score', as this only has missing values
df.drop("User_Score", axis=1, inplace=True)
# Check if everything worked correctly
df.head()

Rank              Name Platform    Year             Genre  Critic_Score  \
0     1            Tetris   Series  1989.0            Puzzle           NaN   
1     2           Pokemon   Series  1998.0      Role-Playing           NaN   
2     3      Call of Duty   Series  2003.0           Shooter           NaN   
3     4       Super Mario   Series  1983.0          Platform           NaN   
4     5  Grand Theft Auto   Series  1998.0  Action-Adventure           NaN   

          Publisher          Developer  NA_Sales  PAL_Sales  JP_Sales  \
0        Nintendo    Alexey Pajitnov         NaN        NaN       NaN   
1        Nintendo         Game Freak         NaN        NaN       NaN   
2      Activision      Infinity Ward         NaN        NaN       NaN   
3        Nintendo           Nintendo         NaN        NaN       NaN   
4  Rockstar Games     Rockstar North         NaN        NaN       NaN   

   Other_Sales  Total_Sales  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN

In [6]:
# Check for duplicates
df.duplicated().sum()

0

In [7]:
# Remove rows that have missing values
df.dropna(axis=0, how="any", inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1718 entries, 25 to 20260
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          1718 non-null   int64  
 1   Name          1718 non-null   object 
 2   Platform      1718 non-null   object 
 3   Year          1718 non-null   float64
 4   Genre         1718 non-null   object 
 5   Critic_Score  1718 non-null   float64
 6   Publisher     1718 non-null   object 
 7   Developer     1718 non-null   object 
 8   NA_Sales      1718 non-null   float64
 9   PAL_Sales     1718 non-null   float64
 10  JP_Sales      1718 non-null   float64
 11  Other_Sales   1718 non-null   float64
 12  Total_Sales   1718 non-null   float64
dtypes: float64(7), int64(1), object(5)
memory usage: 187.9+ KB


## Correct rounding errors

In [8]:
# Check for rounding erros
df["Total_Sales"] != df["NA_Sales"] + df["PAL_Sales"] + df["JP_Sales"] + df[
    "Other_Sales"
]

25        True
58       False
73       False
80       False
91        True
         ...  
17244     True
17714    False
18946     True
19650    False
20260     True
Length: 1718, dtype: bool

In [9]:
# Build own 'Total_Sales' column
df["Total_Sales"] = (
    df["NA_Sales"] + df["PAL_Sales"] + df["JP_Sales"] + df["Other_Sales"]
)
# Check if everything worked correctly
df["Total_Sales"] != df["NA_Sales"] + df["PAL_Sales"] + df["JP_Sales"] + df[
    "Other_Sales"
]

25       False
58       False
73       False
80       False
91       False
         ...  
17244    False
17714    False
18946    False
19650    False
20260    False
Length: 1718, dtype: bool

## Bring dataframe into desired format

In [10]:
# Column 'Rank' is no longer needed for further analysis
df.drop("Rank", axis=1, inplace=True)
# Reset Index
df.reset_index(drop=True, inplace=True)
# Check if everything worked correctly
df.head()

Name Platform    Year     Genre  Critic_Score   Publisher  \
0           Wii Sports      Wii  2006.0    Sports           7.7  Nintendo     
1  Mario Kart 8 Deluxe       NS  2017.0    Racing           9.3  Nintendo     
2    Super Mario Bros.      NES  1985.0  Platform          10.0  Nintendo     
3       Mario Kart Wii      Wii  2008.0    Racing           8.2  Nintendo     
4    Wii Sports Resort      Wii  2009.0    Sports           8.0  Nintendo     

        Developer  NA_Sales  PAL_Sales  JP_Sales  Other_Sales  Total_Sales  
0  Nintendo EAD       41.36      29.02      3.77         8.51        82.66  
1  Nintendo EPD        5.05       4.98      2.11         0.91        13.05  
2  Nintendo EAD       29.08       3.58      6.81         0.77        40.24  
3  Nintendo EAD       15.91      12.92      3.80         3.35        35.98  
4  Nintendo EAD       15.61      10.99      3.29         3.02        32.91

In [11]:
# Convert year column from float to int
df["Year"] = df["Year"].astype("int")

In [12]:
# Final check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1718 entries, 0 to 1717
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          1718 non-null   object 
 1   Platform      1718 non-null   object 
 2   Year          1718 non-null   int32  
 3   Genre         1718 non-null   object 
 4   Critic_Score  1718 non-null   float64
 5   Publisher     1718 non-null   object 
 6   Developer     1718 non-null   object 
 7   NA_Sales      1718 non-null   float64
 8   PAL_Sales     1718 non-null   float64
 9   JP_Sales      1718 non-null   float64
 10  Other_Sales   1718 non-null   float64
 11  Total_Sales   1718 non-null   float64
dtypes: float64(6), int32(1), object(5)
memory usage: 154.5+ KB


In [13]:
# Save file for dashboard
df.to_excel("clean_game_sales.xlsx", sheet_name="data")
# Save file for exploratory data analysis
df.to_csv("clean_game_sales.csv", encoding="utf-8", index=False)

<a id='4'></a>
# Conclusions
After the cleaning process, only 1718 of the 25,000 values remain. However, these do not have any missing values and are ideally suited for further processing.

<a id='5'></a>
# References
1. [Numpy](https://numpy.org/doc/stable/)
2. [Pandas](https://pandas.pydata.org/docs/)